### DISCUSSION TITLE
Js script issue with Cefsharp 96

### DISCUSSION THREAD

**Discussion by Drmsingh** [cefsharp/CefSharp#3943] [General]
JS Script working with the cefsharp 95. But after upgraded with cefsharp 96 getting issue with ExecuteScriptAsync when run custom javascript or call the function through native. 
![image](https://user-images.githubusercontent.com/91657896/148070106-a187267e-38cf-409f-8026-5e67086e09e8.png)

![image](https://user-images.githubusercontent.com/91657896/148071505-25c7785a-ebb4-4ba5-a858-bab8e3a96cee.png)

- **What version of the product are you using?**
   -96.0.180

- **What architecture x86 or x64?**
    x86
	
- **What version of .Net?**
    .Net 4.6.1
    
- **On what operating system?**
    Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    WPF
    
- **What steps will reproduce the problem?**
      TRY_JS_MESSAGE_FORMAT = "try {{ {0} }} catch( error ){{ console.log(error) }};";
       message = String.Format(TRY_JS_MESSAGE_FORMAT, message);
					browser.ExecuteScriptAsync(message);

- **What is the expected output? What do you see instead?**
should not appears error with the javascript and successfully work.

- **Please provide any additional information below.**
    Same error with all 96 cefsharp version.

**Comment by campersau** [comment]
What value does `message` contain?
This for example works for me:
```c#
  var message = "1+1";
  var TRY_JS_MESSAGE_FORMAT = "try {{ {0} }} catch( error ){{ console.log(error) }};";
  message = string.Format(TRY_JS_MESSAGE_FORMAT, message);
  browser.ExecuteScriptAsync(message);
```

**Comment by Drmsingh** [comment]
> What value does `message` contain? This for example works for me:
> 
> ```cs
>   var message = "1+1";
>   var TRY_JS_MESSAGE_FORMAT = "try {{ {0} }} catch( error ){{ console.log(error) }};";
>   message = string.Format(TRY_JS_MESSAGE_FORMAT, message);
>   browser.ExecuteScriptAsync(message);
> ```

Any js function look like below script

message=try { _onForeground() } catch( error ){ console.log(error) };

or

message=try { if (window.onLoad) onLoad("","en_us","","{\"winVersion\":10}") } catch( error ){ console.log(error) };

**Comment by amaitland** [comment]
Run your script directly in DevTools and see if you get the same behaviour. 

https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#javascript-debugging

Converting to discussion as this is unlikely a bug in CefSharp. Most likely a result of a change in Chromium.

**Comment by chendemand** [comment]
I use Cefsharp 95 , use this , chromiumWebBrowser1.ExecuteScriptAsync("document.getElementsByTagName('input')[2].click();"); can work. But use Cefsharp 96 don't work it . 
![This is an image](https://raw.githubusercontent.com/chendemand/img/main/%E6%93%B7%E5%8F%96.PNG)

**Reply by chendemand** [reply]
than i use this  ==>chromiumWebBrowser1.ExecuteScriptAsync("(function(){document.getElementsByTagName('input')[2].click();})"); ==> is ok @Drmsingh

**Reply by amaitland** [reply]
> chromiumWebBrowser1.ExecuteScriptAsync("document.getElementsByTagName('input')[2].click();"); can work. But use Cefsharp 96 don't work it .

Thanks for the example. That is indeed a bug in CefSharp.

**Reply by Drmsingh** [reply]
> than i use this ==>chromiumWebBrowser1.ExecuteScriptAsync("(function(){document.getElementsByTagName('input')[2].click();})"); ==> is ok @Drmsingh

I have tried but not working for me. One more thing, when not passing function, just  passing the js statement before function look  like.. message=try { if _(window.onLoad) onLoad("","en_us","","{"winVersion":10}") } catch( error ){ console.log(error) };_  then?

**Reply by campersau** [reply]
This commit changed the extension method `this` parameter

https://github.com/cefsharp/CefSharp/commit/b72d41353e6e9c964a51ea66c5aaad2527f9a383#diff-365a64085d19bad852c00c0769656b9a2deb678b74ac09807fe102e1f9787b72R501

 so when calling `IWebBrowser.ExecuteScriptAsync` it now chooses this method
https://github.com/cefsharp/CefSharp/blob/604e9570f7147a2aebaed5322d3d437d828794d4/CefSharp/WebBrowserExtensions.cs#L472

which appends `();` as it expects just a method name.

So casting `IWebBrowser` to `IChromiumWebBrowserBase` works:
```c#
((IChromiumWebBrowserBase)browser).ExecuteScriptAsync(message);
```

Possible fix: https://github.com/cefsharp/CefSharp/commit/a6b14b7c4c20c86f7fd1dbbe6918072b14145531

**Reply by amaitland** [reply]
I don't recommend using any such workaround as they will likely break when this is fixed.

Using the following should work as expected and is functionality the same (JavaScript is executed on the main frame by default).

```c#
chromiumWebBrowser.GetMainFrame().ExecuteJavaScriptAsync(script); 
```

**Comment by amaitland** [comment]
Unfortunately I cannot reopen the original issue so #3944 will now be used to track this issue.
